README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# sticRs: The R package for the [STICS](https://www6.paca.inra.fr/stics_eng/) model <img src="man/figures/logo.png" alt="logo" width="150" align="right" />

[![Travis-CI Build
Status](https://travis-ci.com/VEZY/sticRs.svg?branch=master)](https://travis-ci.com/VEZY/sticRs)
[![AppVeyor Build
status](https://ci.appveyor.com/api/projects/status/cu1nyxrhc6nmpt5i/branch/master?svg=true)](https://ci.appveyor.com/project/VEZY/sticrs/branch/master)
[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![License: GPL
v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Coverage
status](https://codecov.io/gh/VEZY/sticRs/branch/master/graph/badge.svg)](https://codecov.io/github/VEZY/sticRs?branch=master)
[![DOI](https://zenodo.org/badge/133052970.svg)](https://zenodo.org/badge/latestdoi/133052970)

## Overview

This package allows the user to programmatically:

  - read ([`read_param`](R/read_param.R)) or set
    ([`set_param`](R/read_param.R)) parameters  

  - set ([`set_usm`](R/set_usm.R)) one or more simulations (USM)

  - call STICS to run them ([`run_stics`](R/run_stics.R))

  - import the results for analyzes ([`read_output`](R/read_output.R)),
    or the observations available in the USM
    ([`read_obs`](R/read_obs.R))

  - compare observations and simulations
    ([`eval_output`](R/eval_output.R))

  - make all previous at once in parallel to evaluate and compare the
    output of STICS versions or parameter values effect
    ([`stics_eval`](R/stics_eval.R))

  - run sensitivity analyzes ([`sensitive_stics`](R/sensitive_stics.R))
    to one or more input parameters and their possible interactions on
    one or more output variables. Note that
    [`stics_eval`](R/stics_eval.R) can evaluate parameter change effect
    also, but doesn’t run full sensitivity analyzes.

  - and optimize parameter values ([`optimi_stics`](R/optimi_stics.R))

The package is under intensive development, so you can fill an issue or
request me a feature [here](https://github.com/VEZY/sticRs/issues) at
any time.

The following diagram presents a typical workflow using the main
functions from `sticRs`:

A scientific article is available
[here](https://github.com/VEZY/sticRs/blob/master/paper/paper.pdf).

## Installation

The development version from [GitHub](https://github.com/) can be
installed with:

``` r
devtools::install_github("VEZY/sticRs")
```

Or using the lightweight
[remotes](https://github.com/r-lib/remotes#readme) package:

``` r
# install.packages("remotes")
remotes::install_github("VEZY/sticRs")
```

The package is tested routinely to pass all
[CRAN](https://CRAN.R-project.org) tests using Travis-CI (linux) and
AppVeyor (Windows), but it is not released to the CRAN servers because
we believe sticRs users are not widespread enough to bother CRAN people
and use their free server time.

## Examples

Toy examples are given here for users in a hurry, but it is highly
recommended to read the [introductory
vignette](https://vezy.github.io/sticRs/articles/Introduction_to_sticRs.html)
for a good start, and the
[sensitivity](https://vezy.github.io/sticRs/articles/Sensitivity_analyses.html)
and [parameter
optimization](https://vezy.github.io/sticRs/articles/optimisation.html)
vignettes if needed.

### Setting a parameter, running the model and compare with observations

#### Manually

This is a basic example using the default dummy simulation (parameters
and meteorology) for a mixed crop of wheat-wheat (not a real mixed crop,
for testing the model behavior) :

``` r
library("sticRs")
# Path to a preconfigured USM:
path_origin_USM= "Your_path_goes_here"
# Path where the simulation will be made:
path_simulation= "Your_path_goes_here_again"
# Path to the STICS model executable:
path_STICS= "Your_path_goes_here_again_and_again"
# Importing the preconfigured USM into a new folder:
import_usm(dir.orig = path_origin_USM, 
           dir.targ = path_simulation,
           stics = path_STICS, 
           usm_name = "test")
# Reading the interrang parameter for both plants (= interrow):
read_param(dirpath = file.path(path_simulation,"test"),
           param='interrang')
# Setting the interrang parameter to 0.30m for both plants:
set_param(dirpath = file.path(path_simulation,"test"),
          param= "interrang", value= 0.3,plant = c(1,2))
# Setting the outputs needed from STICS:
set_out_var(filepath = file.path(path_simulation,"test","var.mod"),
            vars = c("hauteur","lai(n)",'masec(n)'))
# Running the model:
run_stics(dirpath= file.path(path_simulation,"test"))
# Reading the model outputs:
out= read_output(dirpath= file.path(path_simulation,"test"))
# Plotting automatically the outputs along the observations: 
plot_output(file.path(path_simulation,"test"),
            obs_name = c("wheat.obs","pea.obs"))
```

To use your own data, simply use the folder of your simulation as the
reference path; like you would do with javaSTICS.

> To find a possible output variable with a partial match, you can use
> the `find_STICS_var` function.

#### Automatically

You can run all previous code using the simple, standardized
`stics_eval` function:

``` r
library("sticRs")
out= 
  stics_eval(dir.orig = path_origin_USM, 
             dir.targ = path_simulation,
             stics = path_STICS,
             Parameter = list(interrang= 0.3),Plant = c(1,2),
             obs_name = c("wheat.obs","pea.obs"),
             Out_var = c("hauteur","lai(n)",'masec(n)'),
             Title = "Wheat-Wheat", plot_it = T)
```

This function will import the USM in a new folder, change the parameter
values, run the model, return the outputs (simulation output + ggplot
object) and plot it.

### Comparing model simulations

#### Comparing STICS outputs with different parameter values

If you want to compare the effect of different values of one or several
parameters values on several STICS outputs, you can give different
parameter values to the `stics_eval` function. Here is an example with
the `P_rapforme` parameter:

``` r
Eval_stics= 
  stics_eval(dir.orig = path_origin_USM, 
             dir.targ = path_simulation,
             stics = path_STICS,
             Parameter = list(P_rapforme= list(1.5,2,4)),Plant = c(1,2),
             obs_name = c("wheat.obs","pea.obs"),
             Out_var = c("hauteur","laisen(n)","lai(n)","eai","largeur",
                         "varrapforme","dfol","dominant"),
             Title = "Wheat-Wheat", plot_it = T)
```

The function will return a list of STICS outputs for each parameter
value, and a plot comparing the model simulations.

#### Comparing STICS outputs with different STICS versions

If you want to compare different versions of the model after modifying
the code for example, you can give different stics values to the
`stics_eval` function. Here is an example:

``` r
Eval_stics= 
  stics_eval(dir.orig = path_origin_USM, 
             dir.targ = path_simulation,
             stics = list(Original= path_STICS,
                          Modified= "Path_to_modified_stics/stics.exe"),
             obs_name = c("wheat.obs","pea.obs"),
             Out_var = c("hauteur","laisen(n)","lai(n)","eai","largeur",
                         "varrapforme","dfol","dominant"),
             Title = "Wheat-Wheat", plot_it = T)
```

The function will return a list of STICS outputs for each stics
executable provided, and a plot comparing the model simulations.

### Making a sensitivity analysis

Make a sensitivity analysis using the `fast99` algorithm for the
interrow (`interrang` parameter) for three main variables: the
intercepted radiation (`raint`), the leaf area index (`lai(n)`), and the
dry mass (`masec(n)`):

``` r
library("sticRs")
sens= sensitive_stics(dir.orig = path_origin_USM,
                      dir.targ = path_simulation,
                      stics = path_STICS,
                      obs_name = "Wheat.obs",Parameters = "interrang",
                      Vars = c("raint", "lai(n)", "masec(n)"),
                      method= "fast99", n= 10,
                      q= "qunif",Parameters = list(interrang= list(min=0.05, max=0.25),
                                              interrang= list(min=140, max=280)))
```

NB: `n` and `q` are parameters from the
[`fast99`](https://cran.r-project.org/web/packages/sensitivity/sensitivity.pdf)
function, and the `Parameters` argument is passed to the `q.arg`
parameter from `fast99`.

The output from [`sensitive_stics`](R/sensitive_stics.R) is a list of
two:

  - A list of ggplot objects to plot the sensitivity of each variable to
    the parameter(s) along the rotation

  - A list of the output from the method function, *e.g.* a list of
    class `fast99` for the `fast99` method.

## Example data

Example data are available in the [tests
folder](https://github.com/VEZY/sticRs/tree/master/tests/testthat/example_data),
but also as a [separate repository](https://github.com/VEZY/STICS_dummy)
for download convenience. Download instructions are available on the
companion repository.  
This example data is a dummy USM of wheat in self-intercropping, meaning
that the model is run on the same plant planted in intercropping, to
test if the model outputs are close to a sole crop simulation.  
**Warning**: this example USM is made primarily to test the sticRs
package, and is available to the user only for training, not for model
validation. These data are dummy data that were entirely fabricated from
scratch. It is not reflecting any real observations.

## Code of conduct

Please note that this project is released with a [Contributor Code of
Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree
to abide by its terms.

## Authors and acknowledgments

The STICS (Simulateur mulTIdisciplinaire pour les Cultures Standard, or
multidisciplinary simulator for standard crops) model is a dynamic,
generic and robust model aiming to simulate the soil-crop-atmosphere
system. It was first developed in 1996 by INRA -the French National
Institute for Agricultural research- by Nadine Brisson and Dominique
Ripoche. An overview of the model is available
[here](https://www6.paca.inra.fr/stics_eng/About-us/Stics-model-overview).

The sticRs package was developed by [Rémi
Vezy](https://remi-vezy.netlify.com/) and the [STICS
group](https://www6.paca.inra.fr/stics_eng/) thanks to the European
H2020 funded [ReMIX project](https://www.remix-intercrops.eu/).

## ![ReMIX logo](man/figures/remix_logo.jpg)


NEWS.md
# sticRs 1.2.1

Fix issue when using `stics_eval` with more than one parameter

# sticRs 1.2.0

* optimi_stics now accepts further arguments to pass to the optimisation functions (e.g. `tol` for the convergence tolerance)

# sticRs 1.1.3

* Bug fix when using set_param for associated plant

# sticRs 1.1.2

* optimi_stics now can restart automatically with new start values if the optimization process returns an error (e.g. singular matrix)  
* Add possibility to use integer parameter type in optimi_stics

# sticRs 1.1.1

* Fix several bugs in parameter optimization
* Best optimization is now returned by optimi_stics instead of the last one
* Add tests for optimization -> parameters min/max/start values

# sticRs 1.1.0
Add parameter optimization (adapted to intercropping)

# sticRs 1.0.2
Add vignettes for introduction and for sensitivity analyses.

# sticRs 1.0.1

Add exact match to set_* and read_param to avoid issue when using higher level functions such as stics_eval with similar parameter names (e.g. P_hauteur_threshold and P_hauteur_threshold_2).

# sticRs 1.0.0

* Added a `NEWS.md` file to track changes to the package.
* First stable release of the package


R/stics_eval.R
#' Evaluate STICS version
#'
#' @description Wrapper function to run STICS simulations and evaluate
#'  outputs using observations and/or other model versions.
#'
#' @param dir.orig   Path to the directory from which to copy the simulation files. If
#'                   `NULL` (the default), uses the package dummy USM.
#' @param dir.targ   Path to the target directory for evaluation. Created if missing.
#' @param stics      STICS executable path named list (see details).
#' @param Parameter  STICS input parameter named list (see details).
#' @param Plant      Integer value of the plant on which to set the parameter if STICS
#'                   is run on mixed crop (1= Principal, 2= Associated)
#' @param obs_name   A vector of observation file name(s). It must have the form
#'                   `c(Dominant,Dominated)` for mixed crops.
#'                   See [read_obs()] `filename` parameter for more details.
#' @param Out_var    The variables needed as output
#' @param plot_it    Boolean. Do the plot has to be pinted ?
#' @param Parallel   Are the simulations to be executed in parallel ?
#' @param mixed      (optional) Is the simulation made on mixed species (boolean)
#' @param Title      A title for the evaluation. This will be used on the ggplot object.
#' @param Erase      Should the simulations data be erased upon import (see details)?
#'
#' @details The function evaluate STICS outputs either along different model versions \strong{OR}
#' parameter values, not both at the same time. The method is automatically chosen using the
#' `stics` and `Parameter` length. The parameter with a length > 1 will be evaluated.
#' The names of the `stics` or the `Parameter` list are used for reference in the outputs
#' (data.frames and plots). The names are mandatory for `Parameter`, but are optionnal for
#' `stics`. If no names are provided for `stics`, the function give a dummy name for each
#' model evaluation.
#' The format of both `stics` or `Parameter` parameters is the same: a named list of either
#' STICS executable path or a named list of parameters value(s).
#' Please set the `Parameter` argument to `NULL` (the default) for no parameter changes.
#' The function run STICS executable in parrallel using all cores from the machine but one.
#'
#' @return A list of three objects:
#' \describe{
#'   \item{outputs}{A list of `data.frame` objects corresponding
#'   to the simulation for each value in the `stics` or `Parameter` arguments. The
#'   data.frame is made by a call to [eval_output()].}
#'   \item{gg_object}{A summary plot of the simulations outputs returned as a ggplot object.
#'    Possibly a comparison between simulations if several values are given to the `stics`
#'    or `Parameter` arguments.}
#'    \item{stats}{A `data.frame` of a set of summary statistics for each simulation, computed
#'    by a call to [stati_stics()].}
#' }
#'
#'
#' @seealso [sensitive_stics()] to evaluate STICS sensitivity to parameter(s), and other
#' functions used under the hood: [eval_output()], [import_usm()],
#' [run_stics()], [stati_stics()], and [set_out_var()].
#'
#' @examples
#' \dontrun{
#' library(sticRs)
#' # Evaluating a change in a parameter:
#'
#' Eval_parameter=
#'  stics_eval(dir.orig = "0-DATA/dummy/IC_Wheat_Pea",
#'             dir.targ = "2-Simulations/Parameterization",
#'             stics = list(EquivDens= "0-DATA/stics_executable/EquDens/stics.exe"),
#'             Parameter =  list(interrang= c(0.05, 0.25)),
#'             obs_name =  c("6_IC_Wheat_N02.obs","6_IC_Pea_N02.obs"),
#'             Out_var = c("raint", "trg(n)", "rsoleil", "lai(n)", "masec(n)",
#'             "hauteur", "cumraint", "fapar", "eai"),
#'             Title = "Wheat-Pea Auzeville 2005-2006 N0", plot_it = T)
#'
#'
#' # Evaluating a change in the executable:
#' Eval_stics=
#' stics_eval(dir.orig = "0-DATA/dummy/IC_Wheat_Pea",
#'            dir.targ = "2-Simulations/Parameterization",
#'            stics = list(EquivDens= "0-DATA/stics_executable/EquDens/stics.exe"),
#'            obs_name =  c("6_IC_Wheat_N02.obs","6_IC_Pea_N02.obs"),
#'            Out_var = c("raint", "trg(n)", "rsoleil", "lai(n)", "masec(n)",
#'            "hauteur", "cumraint", "fapar", "eai"),
#'            Title = "Wheat-Pea Auzeville 2005-2006 N0", plot_it = T)
#' }
#' @export
stics_eval= function(dir.orig=NULL, dir.targ= getwd(),stics,Parameter=NULL,
                     Plant=1,obs_name= NULL,
                     Out_var=c("lai(n)","masec(n)","hauteur"), plot_it=T,
                     Parallel=T,mixed= NULL,Title=NULL,Erase=TRUE){

  if(!dir.exists(dir.targ)&Erase){erase_dir=T}else{erase_dir=F}
  Param_val= Parameter
  if(length(stics)>1&any(lapply(Param_val, length)%>%unlist > 1)){
    stop("stics_eval only evaluate several STICS executables OR parameter values",
         " not both at a time.", "\nPlease provide only one with several values")
  }

  # Setting usm name to either parameter values or stics names
  if(any(lapply(Param_val, length)%>%unlist > 1)){
    if(is.list(Parameter)){
      usm_name= paste("Set",seq_along(Param_val[[1]]), sep="_")
      method= "Parameter"
    }else{
      stop("The Parameter parameter must be a list")
    }
  }else{
    if(is.list(stics)){
      usm_name= names(stics)
      if(is.null(usm_name)){
        usm_name= paste0("STICS_",seq_along(stics))
      }
      if(length(stics)>1){
        method= "stics"
      }else{
        method= "Parameter" # Default method
      }
    }else{
      stop("The stics parameter must be a list")
    }
  }

  # Add standard exe name if needed:
  stics= lapply(stics, function(x){
    if(file.info(x)$exe == "no"){
      file.path(x,"stics")
    }else{
      x
    }
  })

  if(method == "Parameter"){
    stics= rep(stics, length(usm_name))
  }

  if(Parallel&length(usm_name)>1){
    NbCores= parallel::detectCores()-1
    cl= parallel::makeCluster(min(NbCores,length(usm_name)))
    parallel::clusterExport(cl=cl,
                            varlist=c("dir.orig","dir.targ","usm_name","stics",
                                      "obs_name","Out_var","import_usm",
                                      "set_out_var","Plant","run_stics",
                                      "eval_output","Parameter","Erase","method",
                                      "set_param","Param_val"),
                            envir=environment())
    # The apply function runs either along parameter values or stics exe.
    outputs=
      parallel::parLapply(
        cl,
        seq_along(usm_name),
        function(x,dir.orig,dir.targ,usm_name,stics,
                 obs_name,Parameter,Plant,Erase,method,
                 Param_val){
          tryCatch(expr = {
            USM_path= file.path(dir.targ,usm_name[x])
            import_usm(dir.orig = dir.orig, dir.targ = dir.targ,
                       usm_name = usm_name[x], overwrite = T,
                       stics = ifelse(method=="stics",stics[[x]],
                                      stics[[1]]))
            set_out_var(filepath= file.path(USM_path,"var.mod"),
                        vars=Out_var, add=F)
            if(method=="Parameter"){
              # set parameter values using mapply:
              Param_val_x= lapply(Param_val, function(y){y[x]})
              mapply(function(y,z){
                set_param(dirpath = USM_path,
                          param = y,plant = Plant,
                          value = z)
              },names(Param_val_x),Param_val_x)
            }
            run_stics(dirpath = USM_path,basename(stics[[x]]))
            output= eval_output(dirpath= USM_path, obs_name= obs_name)
            if(Erase){
              unlink(x = USM_path, recursive = T, force = T)
            }
            output
          },
          error=function(cond) {
            message(paste("Error during simulation setup for USM:",usm_name[x]))
            message("Here's the original error message:")
            message(cond)
            unlink(x = USM_path, recursive = T, force = T)
            return(NULL)
          },
          warning=function(cond) {
            message(paste("Simulation setup caused a warning for USM:",usm_name[x]))
            message("Here's the original warning message:")
            message(cond)
            return(NULL)
          },
          finally={
            if(Erase){
              unlink(x = USM_path, recursive = T, force = T)
            }
          }
          )
        },dir.orig,dir.targ,usm_name,stics,obs_name,
        Parameter,Plant,Erase,method,Param_val)
    parallel::stopCluster(cl)
  }else{
    outputs=
      lapply(seq_along(usm_name),
             function(x){
               USM_path= file.path(dir.targ,usm_name[x])
               import_usm(dir.orig = dir.orig, dir.targ = dir.targ,
                          usm_name = usm_name[x], overwrite = T,
                          stics = ifelse(method=="stics",
                                         stics[[x]],stics[[1]]))
               set_out_var(filepath= file.path(USM_path,"var.mod"),
                           vars=Out_var, add=F)
               if(method=="Parameter"){
                 # set parameter values using mapply:
                 Param_val_x= lapply(Param_val, function(y){y[x]})
                 mapply(function(y,z){
                   set_param(dirpath = USM_path,
                             param = y,plant = Plant,
                             value = z)
                 },names(Param_val_x),Param_val_x)
               }
               run_stics(dirpath = USM_path,basename(stics[[x]]))
               output= eval_output(dirpath= USM_path, obs_name= obs_name)
               if(Erase){
                 unlink(x = USM_path, recursive = T, force = T)
               }
               output
             })
  }

  if(erase_dir){
    unlink(x = dir.targ, recursive = T, force = T)
  }

  names(outputs)= usm_name
  stats_out= do.call(function(...)try(stati_stics(...),silent = T),outputs)
  if(inherits(stats_out,"try-error")){
    warning("Can't find ",crayon::red("ANY"),
            " valid observation to perform evaluation")
  }
  outputs[["plot_it"]]= plot_it
  outputs[["Vars"]]= Out_var
  outputs[["Title"]]= Title
  gg_stics= do.call(plot_output,outputs)

  return(list(outputs= outputs[-grep("plot_it|Vars|Title",names(outputs))],
              gg_object= gg_stics,
              stats= stats_out))
}


#' Makes one stics_eval simulation list from several
#'
#' @description Uses [data.table::rbindlist()] but keep the original
#' structure of the list within the list. To use for row binding
#' [stics_eval()] outputs.
#'
#' @param ...   Input simulations lists from [stics_eval()].
#'
#' @details The function keep all columns from the simulations as in `data.table::rbindlist(l,fill= TRUE)`.
#'
#' @seealso [stics_eval()]
#'
#' @importFrom data.table rbindlist
#'
#' @examples
#' \dontrun{
#'
#' Simulation= rbind_sim(Sim_2017, Sim_2018)
#'
#' }
#' @export
#'
rbind_sim= function(...){
  dot_args= list(...)
  bind_s=
    lapply(names(dot_args[[1]]$outputs), function(x){
      data.table::rbindlist(
        lapply(dot_args, function(dot_x,y,x){dot_x[[y]][[x]]}, "outputs",x),
        fill = TRUE)%>%
        as.data.frame()
    })
  names(bind_s)= names(dot_args[[1]]$outputs)
  return(bind_s)
}

